packages/openid4vc-client/tests/openid4vc-client.e2e.test.ts

```diff
@@ -98,6 +98,137 @@ describe('OpenId4VcClient', () => {
         'OpenBadgeCredential',
       ])
 
+      // @ts-ignore
+      expect(w3cCredentialRecord.credential.credentialSubject.id).toEqual(did.didState.did)
+    })
+  })
+  describe('Authorization flow', () => {
+    beforeAll(async () => {
+      /**
+       *  Below we're setting up some mock HTTP responses.
+       *  These responses are based on the openid-initiate-issuance URI above
+       * */
+
+      // setup temporary redirect mock
+      nock('https://launchpad.mattrlabs.com').get('/.well-known/openid-credential-issuer').reply(307, undefined, {
+        Location: 'https://launchpad.vii.electron.mattrlabs.io/.well-known/openid-credential-issuer',
+      })
+
+      // setup server metadata response
+      const httpMock = nock('https://launchpad.vii.electron.mattrlabs.io')
+        .get('/.well-known/openid-credential-issuer')
+        .reply(200, getMetadataResponse)
+
+      // setup access token response
+      httpMock.post('/oidc/v1/auth/token').reply(200, acquireAccessTokenResponse)
+
+      // setup credential request response
+      httpMock.post('/oidc/v1/auth/credential').reply(200, credentialRequestResponse)
+    })
+
+    afterAll(async () => {
+      cleanAll()
+      enableNetConnect()
+    })
+
+    it('should generate a correct code verifier', async () => {
+      const codeVerifier = await agent.modules.openId4VcClient.generateCodeVerifier()
+
+      expect(typeof codeVerifier).toBe('string')
+
+      // length restrictions are specified here https://www.rfc-editor.org/rfc/rfc7636#section-4.1
+      expect(codeVerifier.length).toBeGreaterThanOrEqual(43)
+      expect(codeVerifier.length).toBeLessThanOrEqual(128)
+    })
+
+    it('should generate a valid authorization url', async () => {
+      const clientId = 'test-client'
+      const codeVerifier = await agent.modules.openId4VcClient.generateCodeVerifier()
```

Comment:
Do we need to expose this as a separate method? Do you think we could generate it in the generateAurhorizationUrl and just return it as part of the method?

For all other methods in AFJ we also just generate the nonce

Reply:
The caller needs to pass the same `codeVerifier` value when generating the url (`generateAuthorizationUrl`) as when requesting the credential (`requestCredential`) for the flow to succeed. This is to prevent others from intercepting the authorization code and pretending they're us. If we don't expose this method, we'd have to store and manage it internally. That seemed a bit extra to me :)

Reply:
We could just generate it and return it in the generateAuthorizationUrl?